File: PRE/geometry.cpp
/*
 * The Point<T> and Line<T> classes (where T is the scalar type), along with
 * several 2D geometry algorithms.
 */

/*
 * Most of the arithmetic operations here use the usual notation; additionally,
 * (*) and (%) refer to the dot and cross product.
 */
template <typename T>
struct Point {
T x, y;
inline Point & operator+=(const Point & o) { x += o.x, y += o.y; return *this; }
inline Point operator+(const Point & o) const { auto p = *this; return p += o; }
inline Point & operator-=(const Point & o) { x -= o.x, y -= o.y; return *this; }
inline Point operator-(const Point & o) const { auto p = *this; return p -= o; }
inline T operator*(const Point & o) { return x * o.x + y * o.y; }
inline T operator%(const Point & o) const { return x * o.y - y * o.x; }
inline bool operator==(const Point & o) const { return x == o.x && y == o.y; }
inline bool operator<(const Point & o) const { return x < o.x || (x == o.x && y < o.y); }
inline double r() const { return hypot(x, y); }
inline T q() const { return atan2(y, x); }
};
template <typename T> inline Point<T> operator*(T c, Point<T> a) { return {c * a.x, c * a.y}; }
template <typename T> inline Point<T> operator-(Point<T> a) { return (T) -1 * a; }

/*
 * Calculates the area of a polygon using cross products. The polygon may be
 * convex or concave, but cannot intersect itself.
 */
template <typename It>
auto polygon_area(It p0, It p1) -> decltype(p0->x) {
	int n = p1 - p0;
	decltype(p0->x) ans = 0;
	rep (i, n) ans += p0[i] % p0[(i + 1) % n];
	return ans / 2;
}

/*
 * Computes the vertices of the convex hull of a set of points using Graham's
 * scan. By default, points are returned in counterclockwise order, starting
 * from the lowest leftmost point. Duplicated vertices or points lying on the
 * edge (but which are not vertices) are not included. Returns the number of
 * vertices found.
 */
template <typename IIt, typename OIt>
int convex_hull(IIt p0, IIt p1, OIt q0) {
	auto base = *min_element(p0, p1);
	using P = decltype(base);
	sort(p0, p1, [base](const P & p_, const P & q_) {
		P p = p_ - base, q = q_ - base;
		auto c = p % q;
		return c ? c > 0 : p * p < q * q;
	});
	OIt q1 = q0;
	while (p0 != p1) {
		while (q1 - q0 > 1 && (*(q1 - 1) - *(q1 - 2)) % (*p0 - *(q1 - 2)) <= 0)
			q1--;
		*(q1++) = *(p0++);
	}
	if (q1 - q0 == 2 && q0[0] == q0[1]) q1--;
	return q1 - q0;
}

/*
 * Finds the distance between the closest pair of points in a set of points.
 */
template <typename It>
double closest(It p0, It p1) {
	if (p1 - p0 < 2) return 0;
	using P = typename decay<decltype(*p0)>::type;
	sort(p0, p1, [](const P & p, const P & q) {
		return p.y < q.y || (p.y == q.y && p.x < q.x);
	});
	double ans = (p0[1] - p0[0]).r();
	set<P> w;
	range (p, p0, p1) {
		while (p0->y < p->y - ans)
			w.erase(*(p0++));
		// For integral scalars:
		// auto q = w.lower_bound(*p - P{ceil(ans), ceil(ans)});
		// For floating-point scalars:
		// auto q = w.lower_bound(*p - P{ans, ans});
		while (q != w.end() && q->x <= p->x + ans)
			ans = min(ans, (*p - *(q++)).r());
		w.insert(*p);
	}
	return ans;
}

/*
 * A simple line (segment) class that detects line intersections.
 */
template <typename T>
inline int sgn(T x) { return x < 0 ? -1 : x > 0 ? 1 : 0; }

template <typename T>
struct Line {
	Point<T> a, b;
	inline bool intersects(const Line & o) const { return crosses(o) && o.crosses(*this); }

	private:
	inline bool crosses(const Line & o) const { return sgn((o.a - a) % (b - a)) * sgn((o.b - a) % (b - a)) <= 0; }
};

File: PRE/math/Geometry.java
	static double dist( double[] x1, double[] x2 )
	{
		double ans = 0;
		for( int i = 0; i < x1.length; i++ ) ans += (x1[i]-x2[i])*(x1[i]-x2[i]);
		return Math.sqrt(ans);
	}

	double cross( double x0, double y0, double x1, double y1 )
	{
		return x0 * y1 - x1 * y0;
	}

	boolean intersects( double x0, double y0, double x1, double y1, double x2, double y2, double x3, double y3 )
	{
		double a = cross( x1 - x0, y1 - y0, x2 - x0, y2 - y0 );
		double b = cross( x1 - x0, y1 - y0, x3 - x0, y3 - y0 );
		double c = cross( x3 - x2, y3 - y2, x0 - x2, y0 - y2 );
		double d = cross( x3 - x2, y3 - y2, x1 - x2, y1 - y2 );
		return ( ( a >= 0 && b <= 0 ) || ( a <= 0 && b >= 0 ) ) && ( ( c >= 0 && d <= 0 ) || ( c <= 0 && d >= 0 ) );
	}

	/*
		implements Graham scan for a set of points;
		outputs points on convex hull in counterclockwise order,
		starting with the point in the leftmost, lowest point, e.g.,
		the point with lowest y coord, then lowest x coord
	*/

	static ArrayList<double[]> convex_hull( double[][] pts )
	{
		int n = pts.length;
		// find starting pt
		double[] start = pts[0];
		for( int i = 0; i < n; i++ )
			if( (pts[i][1]<start[1]) || (pts[i][1]==start[1]&&pts[i][0]<start[0]) )
				start = pts[i];
		// sort by angle, then by distance, against starting point
		Integer[] ptr = new Integer[n+1];
		for( int i = 0; i < n; i++ ) ptr[i] = i;
		double[] a = new double[n], d = new double[n];
		for( int i = 0; i < n; i++ )
		{
			a[i] = Math.atan2( pts[i][1] - start[1], pts[i][0] - start[0] );
			d[i] = Math.hypot( start[0] - pts[i][0], start[1] - pts[i][1] );
		}
		Arrays.sort( ptr, 0, n, new AngleComparator( a, d ) );
		ptr[n] = ptr[0];
		// check counterclockwise turning ( turn(a,b,c) > 1 )
		ArrayList<double[]> h = new ArrayList<double[]>();
		h.add(pts[ptr[0]]);
		h.add(pts[ptr[1]]);
		int last = 1;
		for( int i = 2; i <= n; i++ )
		{
			while( true )
			{
				double t = turn( h.get(last-1), h.get(last), pts[ptr[i]] );
				if( t<0 || (t*t)<EPS )
				{
					h.remove(last);
					last--;
				}
				else break;
				if( last==0 ) break;
			}
			last++;
			h.add(pts[ptr[i]]);
		}
		h.remove(last);
		// last case: if the entire thing was just a line
		if( h.size()==1 ) h.add(pts[ptr[n-1]]);
		return h;
	}

	static class AngleComparator implements Comparator<Integer>
	{
		double[] a, d;
		AngleComparator( double[] _a, double[] _d )
		{ a = _a; d = _d; }
		public int compare( Integer x, Integer y )
		{
			if( a[x] > a[y] ) return 1;
			if( a[x] < a[y] ) return -1;
			if( d[x] > d[y] ) return 1;
			if( d[x] < d[y] ) return -1;
			return 0;
		}
	}

	static double turn( double[] a, double[] b, double[] c )
	{ return (b[0]-a[0])*(c[1]-a[1])-(b[1]-a[1])*(c[0]-a[0]); }


/*
	Calculates the area and centroid of the polygon given by x, y.
*/

	static double polygon_area( double[] x, double[] y )
	{
		double a = 0;
		for( int i = 1; i < x.length; i++ ) a += x[i-1]*y[i] - x[i]*y[i-1];
		a += x[x.length-1]*y[0] - x[0]*y[x.length-1];
		a /= 2;
		if( a < 0 ) a = -a;
		return a;
	}

	static double[] polygon_area_centroid( double[] x, double[] y )
	{
		double[] a = new double[3];
		int n = x.length;
		double cross;
		for( int i = 1; i < x.length; i++ )
		{
			cross = x[i-1]*y[i] - x[i]*y[i-1];
			a[0] += cross;
			a[1] += cross*(x[i-1]+x[i]); a[2] += cross*(y[i-1]+y[i]);
		}
		cross = x[n-1]*y[0] - x[0]*y[n-1];
		a[0] += cross;
		a[0] /= 2;
		if( a[0] < 0 ) a[0] *= -1;
		a[1] += cross*(x[n]+x[0]); a[2] += cross*(y[n]+y[0]);
		a[1] /= 6*a[0]; a[2] /= 6*a[0];

		return a;
	}
